src/lib/storage/buffer/frame.cpp

```diff
@@ -0,0 +1,180 @@
+      "Frame must be UNLOCKED to transition to MARKED, instead: " + std::to_string(state(_state_and_version.load())));
+  auto state_and_version = old_state_and_version;
+  return _state_and_version.compare_exchange_strong(state_and_version,
+                                                    _update_state_with_same_version(old_state_and_version, MARKED));
+}
+
+bool Frame::is_dirty() const {
+  return (_state_and_version.load() & _dirty_mask) >> _dirty_shift;
+}
+
+Frame::StateVersionType Frame::state(const Frame::StateVersionType state_and_version) {
+  return (state_and_version & _state_mask) >> _state_shift;
+}
+
+Frame::StateVersionType Frame::version(const Frame::StateVersionType state_and_version) {
+  return state_and_version & _version_mask;
+}
+
+Frame::StateVersionType Frame::state_and_version() const {
+  return _state_and_version.load();
+}
+
+NodeID Frame::node_id() const {
+  return node_id(_state_and_version.load());
+}
+
+NodeID Frame::node_id(const Frame::StateVersionType state_and_version) {
+  return static_cast<NodeID>((state_and_version & _node_id_mask) >> _node_id_shift);
+}
+
+bool Frame::try_lock_shared(const Frame::StateVersionType old_state_and_version) {
+  auto old_state = state(old_state_and_version);
+  auto state_and_version = old_state_and_version;
+
+  // Multiple threads can try to lock shared concurrently until the state reaches MAX_LOCKED_SHARED
+  if (old_state < MAX_LOCKED_SHARED) {
+    // Increment the state by 1 to add a new concurrent reader
+    return _state_and_version.compare_exchange_strong(
+        state_and_version, _update_state_with_same_version(old_state_and_version, old_state + 1));
+  }
+
+  // If the state is MARKED, we just set the frame to 1 reader.
+  if (old_state == MARKED) {
+    return _state_and_version.compare_exchange_strong(
+        state_and_version, _update_state_with_same_version(old_state_and_version, SINGLE_LOCKED_SHARED));
+  }
+  return false;
+}
+
+bool Frame::try_lock_exclusive(const Frame::StateVersionType old_state_and_version) {
+  DebugAssert(state(old_state_and_version) == UNLOCKED || state(old_state_and_version) == MARKED ||
+                  state(old_state_and_version) == EVICTED,
+              "Frame must be unlocked, marked or evicted to lock exclusive, instead: " +
+                  std::to_string(state(old_state_and_version)));
+  auto state_and_version = old_state_and_version;
+  return _state_and_version.compare_exchange_strong(state_and_version,
+                                                    _update_state_with_same_version(old_state_and_version, LOCKED));
+}
+
+bool Frame::unlock_shared() {
+  while (true) {
+    auto old_state_and_version = _state_and_version.load();
+    const auto old_state = state(old_state_and_version);
+    DebugAssert(old_state > 0 && old_state <= MAX_LOCKED_SHARED, "Frame must be locked shared to unlock shared.");
+    // Decrement the state by 1 to remove a concurrent reader
```

Comment:
That lowers the version, right? Say so?

Reply:
No, it des not. Only the state

Reply:
Ah ... misread. If we would decrement a `old_state_and_version`, that would decrement the version?